src/js/utils/element-utils.js

```diff
@@ -143,22 +143,55 @@ export function distance(p1, p2) {
   return Math.sqrt(Math.pow(p2.x - p1.x, 2) + Math.pow(p2.y - p1.y, 2));
 }
 
+const rectCache = new WeakMap();
+
+/**
+ * Get the cached bounding client rect of an element with a configurable cache expiration time.
+ * Animations at 12fps are smooth enough for human perception.
+ * @param {Element} element
+ * @param {number} [time = 1000 / 12]
+ * @returns {DOMRect}
+ */
+export function cachedBoundingClientRect(element, time = 1000 / 12) {
```

Comment:
question(blocking): At least for now, it looks like this is only used by `MediaChromeRange` and its descendants. Can we just make this a method on `MediaChromeRange` with the state owned in its instance to avoid adding additional transient state to modules, given the fights we've been having with memory management/leaks/GC?

Reply:
valid call, if it was only used by the instances of the ranges I'd change it but it's used in both those classes by different internal elements and I see it easily being used by other UI components where there's many repetitive calls to getBoundingClientRect. it seems a better fit for a utility. 

The WeakMap makes sure the entries get cleared when the element is destroys so that seems safe

Reply:
Just confirmed that `getBoundingClientRect()` is not [live](https://html.spec.whatwg.org/multipage/infrastructure.html#live) (See: https://drafts.csswg.org/cssom-view/#dom-element-getboundingclientrect), so it's probably safe to assume that a browser implementation would still hold onto any references "under the hood", unlike cases where we e.g. have function closure capture references. That said, relying on the `WeakMap` makes it non-trivially harder (in my experience) to investigate if/when a memory leak _**does**_ occur (elsewhere). Were we running into noticeable perf issues by invoking `getBoundingClientRect()`? Is this needed/could it be solved in the code where we're using `getBoundingClientRect()` instead? (aka is this needed at all?)

Either way, I won't press too hard on this one, but it would be good to get input from other folks to get their thoughts on this.

Reply:
we call `getBoundingClientRect` on several elements at 60fps and it causes browser layout / reflow
https://gist.github.com/paulirish/5d52fb081b3570c81e3a

that's why I opted for a cached function result with eviction around 12fps